proposals/v3/api.md

```diff
@@ -0,0 +1,815 @@
+
+> **Open Question**:
+> For the 'context' field, described above, do we want it at an individual reading
+> level, at the level of the entire reading response, or for it to be possible at
+> both (e.g. ctx for this reading value in particular vs. ctx that applies to all
+> the readings)
+
+
+### Query Parameters
+
+| Key | Description |
+| --- | ----------- |
+| namespace | The default namespace to use for the specified labels. (default: `default`) |
+| ns | Same as `namespace`, above. |
+| tags | The [tags](tags.md) to filter devices on. If specifying multiple tags, they should be comma-separated. |
+
+### Response
+It is recommended to read the document on [v3 reads](reads.md), as it explains in detail
+the changes to the read response.
+
+#### OK (200)
+```json
+{
+  "a72cs6519ee675b": [
+    {
+      "kind": "temperature",
+      "type": "temperature",
+      "value": 20.3,
+      "timestamp": "2018-02-01T13:47:40.395939895Z",
+      "info": "",
+      "unit": {
+        "symbol": "C",
+        "name": "degrees celsius"
+      },
+      "context": {
+        "host": "127.0.0.1",
+        "sample_rate": 8
+      }
+    }
+  ],
+  "929b923de65a811": [
+    {
+      "kind": "led",
+      "type": "state",
+      "value": "off",
+      "timestamp": "2018-02-01T13:47:40.395939895Z",
+      "info": "",
+      "unit": null
+    },
+    {
+      "kind": "led",
+      "type": "color",
+      "value": "000000",
+      "timestamp": "2018-02-01T13:47:40.395939895Z",
+      "info": "",
+      "unit": null
+    }
+  ],
+  "12bb12c1f86a86e": [
+    {
+      "kind": "door_lock",
+      "type": "lock",
+      "value": "locked",
+      "timestamp": "2018-02-01T13:47:40.395939895Z",
+      "info": "Zone 6B",
+      "unit": null,
+      "context": {
+        "wedge": 1
+      }
+    }
+  ]
+}
+```
+
+#### Error (500, 400)
+See: [Errors](#errors)
+
+
+
+## Read Device
+```
+GET http://HOST:5000/synse/v3/read/{device_id}
+```
+
+Read from the specified device. This endpoint is effectively the same as using the [`/read`](#read)
+endpoint where the label matches the [device id tag](tags.md#auto-generated), e.g.
+`http://HOST:5000/synse/v3/read?tags=id:b33f7ac0`.
+
+### URI Parameters
+
+| Parameter | Description |
+| --------- | ----------- |
+| *device_id* | The ID of the device to read. IDs should be globally unique. |
+
+
+### Response
+
+#### OK (200)
+The response here is similar to the response for the `/read` endpoint, except it just returns
+the list of readings for the device, instead of a collection of 1 `device_id` keyed to this
+device reading info. 
+
+> **Question**: We have a bunch of different response formats for reads - "read", "read device", 
+> and "read cached". They all convey the same data, just structured a bit differently to make it
+> convenient for each of their use cases (bulk read, single read, streamed read -- respectively).
+> Would it make sense to try and unify the response data structures into a more uniform response
```

Comment:
I would prefer to have a somewhat uniform response across all read. But then, for the 3 use cases below, the examples look fine to me.

Reply:
I was thinking about this some more today, mostly around unifying `/read` and `/read/{device_id}`. Depending on how this discussion goes, I can think more about updating the readcached to be something similar.

As a refresher, the `/read` response example from above:
```json
{
  "a72cs6519ee675b": [
    {
      "kind": "temperature",
      "type": "temperature",
      "value": 20.3,
      "timestamp": "2018-02-01T13:47:40.395939895Z",
      "info": "",
      "unit": {
        "symbol": "C",
        "name": "degrees celsius"
      },
      "context": {
        "host": "127.0.0.1",
        "sample_rate": 8
      }
    }
  ],
  "929b923de65a811": [
    {
      "kind": "led",
      "type": "state",
      "value": "off",
      "timestamp": "2018-02-01T13:47:40.395939895Z",
      "info": "",
      "unit": null
    },
    {
      "kind": "led",
      "type": "color",
      "value": "000000",
      "timestamp": "2018-02-01T13:47:40.395939895Z",
      "info": "",
      "unit": null
    }
  ],
  "12bb12c1f86a86e": [
    {
      "kind": "door_lock",
      "type": "lock",
      "value": "locked",
      "timestamp": "2018-02-01T13:47:40.395939895Z",
      "info": "Zone 6B",
      "unit": null,
      "context": {
        "wedge": 1
      }
    }
  ]
}
```

One thought in making things more uniform is to make this a list and move the device_id into a field in the individual read responses, e.g.

```json
[
    {
      "device": "a72cs6519ee675b",
      "kind": "temperature",
      "type": "temperature",
      "value": 20.3,
      "timestamp": "2018-02-01T13:47:40.395939895Z",
      "info": "",
      "unit": {
        "symbol": "C",
        "name": "degrees celsius"
      },
      "context": {
        "host": "127.0.0.1",
        "sample_rate": 8
      }
    },
    {
      "device": "929b923de65a811",
      "kind": "led",
      "type": "state",
      "value": "off",
      "timestamp": "2018-02-01T13:47:40.395939895Z",
      "info": "",
      "unit": null
    },
    {
      "device": "929b923de65a811",
      "kind": "led",
      "type": "color",
      "value": "000000",
      "timestamp": "2018-02-01T13:47:40.395939895Z",
      "info": "",
      "unit": null
    },
    {
      "device": "12bb12c1f86a86e",
      "kind": "door_lock",
      "type": "lock",
      "value": "locked",
      "timestamp": "2018-02-01T13:47:40.395939895Z",
      "info": "Zone 6B",
      "unit": null,
      "context": {
        "wedge": 1
      }
    }
]
```

This pattern can be applied to single device reads as well. The one possibly big caveat here is that by putting the device id into the reading object, the readings themselves are no longer grouped by device, so if a user wanted to get the readings for each device, they would first have to aggregate them on their own.


The alternative would be to format the `/read/{device_id}` in the same manner as the `/read` response. It contains more redundant pieces and nests things a bit more, but its another option.

From a modeling perspective, I think having the device_id in the reading object is a bit nicer because all of the information for a reading is in a single object. With it as the key to the reading, it is kept separate from the actual reading data. In the use case of bulk reads, its sorta annoying to have them separate I think. At least, it wouldn't buy much to do device lookups in the response dict because when doing a bulk operation, they won't necessarily have the device IDs to begin with, so they would just be iterating through all of the readings anyways.

This is getting rambly so I'll stop here, but let me know your thoughts and if you have any ideas on alternatives.

Reply:
the fields that change depending on kind (e.g. `info`, `unit`, `context`) are optional fields which all device readings *can* specify, but do not need to. 

Internally, it wouldn't be hard to omit the empty/null fields if we don't want to propagate it up in the JSON response. This basically just means that whatever is parsing the response needs to know that a missing field is implicitly an empty field.

It would be difficult to do this by checking the device kind/type. The types should be fairly well known (led, temperature, humidity, ...), but different plugins could supply different amounts of info for a reading. Even within the same plugin this could happen (e.g. some temperature sensors are given an "info" in their config while others are not). Checking on the kind is more difficult because they are defined by the plugin. If a new plugin came along that defined a new kind, the only way something upstream would know whether fields are expected or not would be to either read the plugin docs/source, or to have the plugin expose that information to Synse Server somewhere.